pyccel/codegen/printing/ccode.py

```diff
@@ -1649,7 +1652,13 @@ def _print_Allocate(self, expr):
         free_code = ''
         variable = expr.variable
         if isinstance(variable.class_type, (HomogeneousListType, HomogeneousSetType, DictType)):
-            return ''
+            if len(variable.alloc_shape) > 0 and isinstance(variable.class_type, HomogeneousListType):
```

Comment:
Is list the only type with a reserve function? I wouldn't be surprised if one was also available for set and dict

Reply:
Definitely not, I didn't give much thought to this part as i didn't consider all the cases that would require a reserve. I will think more about it

Reply:
I don't know what cases might need reserve either. I just don't see any particular reason to exclude set/dict if they have an equivalent method with a similar name. If they don't then this question becomes irrelevant

Reply:
I excluded them because tests relative to variable annotations were failing.
```
if __name__ == '__main__':
    a : 'set[float]'
    a = {1.5, 2.5, 3.3, 4.3}
```
```
int main()
{
    hset_double a;
    hset_double_reserve(&a, INT64_C(4));
    a = c_init(hset_double, {1.5, 2.5, 3.3, 4.3});
    hset_double_drop(&a);
    return 0;
}
```
this would cause a double free because `c_init` doesn't work well with `reserve` in contrast to `append`. an `Allocate` might need to be treated differently depending on what will happen to the memory.

Reply:
Is this not also true for:
```python
a = [1,2,3,4]
```
?
How do you avoid the reserve/init problem in this case?

Reply:
I think we should update `_assign_lhs_variable` to not add an `Allocate` node when the `rhs` is a `c_init` since it takes care of memory allocation.

Reply:
I'm not sure how you manage this. The semantic stage is language agnostic.  What happens if the Allocate node is needed for Fortran? You put it back in `FCodePrinter._print_PythonList`?

It would be nice if there was a way to make the init value available. Maybe we could add a property to `Allocate`?

Additionally if you have two c_init statements in a row then you will definitely still need the Allocate to know that you need to call c_drop first. (I assume if you can't call c_init on an initialised list then you can't call it twice without some kind of deallocation either).

Eg:
```python
a = [1, 2, 3, 4]
...
a = [6, 7, 8, 9,  10]
```
If `a` was an array instead of a list then this would give:
```fortran
allocate(a(0:3))
a(:) = (/ 1, 2, 3, 4 /)
deallocate(a)
allocate(a(0:4))
a(:) = (/ 6, 7, 8, 9 /)
```
The deallocation comes from the Allocate node